hypertrace-trace-enricher/trace-reader/src/main/java/org/hypertrace/trace/reader/entities/DefaultTraceEntityReader.java

```diff
@@ -133,16 +133,32 @@
         attributes.mapOptional(map -> this.extractString(map, entityType.getNameAttributeKey()));
 
     return zip(
-        id,
-        name,
-        attributes,
-        (resolvedId, resolvedName, resolvedAttributeMap) ->
-            Entity.newBuilder()
-                .setEntityId(resolvedId)
-                .setEntityType(entityType.getName())
-                .setEntityName(resolvedName)
-                .putAllAttributes(resolvedAttributeMap)
-                .build());
+            id,
+            name,
+            attributes,
+            (resolvedId, resolvedName, resolvedAttributeMap) ->
+                Entity.newBuilder()
+                    .setEntityId(resolvedId)
+                    .setEntityType(entityType.getName())
+                    .setEntityName(resolvedName)
+                    .putAllAttributes(resolvedAttributeMap)
+                    .build())
+        .filter(entity -> this.canCreateEntity(entityType, entity));
+  }
+
+  private boolean canCreateEntity(EntityType entityType, Entity entity) {
+    return entityType.getRequiredConditionsList().stream()
+        .allMatch(condition -> this.passesFormationCondition(entity, condition));
+  }
+
+  private boolean passesFormationCondition(Entity entity, EntityFormationCondition condition) {
+    switch (condition.getConditionCase()) {
+      case REQUIRED_KEY:
+        return entity.getAttributesMap().containsKey(condition.getRequiredKey());
+      case CONDITION_NOT_SET:
+      default:
+        return true;
```

Comment:
true or false? 🤔

Reply:
My thinking here was that if no condition was set (or it was unrecognized) we should treat it like it's not there - which is `true` since this is a filter predicate. What do you think?

Reply:
A bit confused. 

If the condition is not set or unrecognized, it would be creating an entity, right? Since, the predicate would return true always in that case, and therefore the `allMatch` as well, where the predicate is being used

Reply:
That's right. Think about it as a filter - if I sent a query with a filter that was not set, I would expect it to return the same as if no filter (which is generally how proto handles default values, too). I could see the argument that the default case (unrecognized) should return false and log an error though - I'll make that change.